File: jobs/src/main/scala/com/hxm/earlgrey/jobs/Job.scala
package com.hxm.earlgrey.jobs

import org.bson.BsonInvalidOperationException
import org.mongodb.scala.Document

/**
  * Created by hxm on 17-4-11.
  */
/*case class Job(id: String, `type`: String, name: String, status: String,
                startedTime: Option[String] = None, endedTime: Option[String] = None,
                params: Option[Map[String, String]] = None, progress: Int = 0,
                statusDetail: Option[String] = None)*/

case class Job(doc: Document) {
  def getParameter(paramName: String): Option[String] = {
    try {
      doc.get(Job.ParamsString).map(_.asDocument.getString(paramName).getValue)
    }
    catch {
      //no key
      case e: BsonInvalidOperationException => None
    }
  }

  def getStatus(): String = doc.getString(Job.StatusString)

}

object Job {
  val TimeFormat = "yyyy-MM-dd-HH:mm:ss"

  object Type {
    val Test = "Test"
    val FlowQuery = "FlowQuery"
  }

  object Status {
    val Created = "Created"
    val Started = "Started"
    val End = "End"
    val Error = "Error"
  }

  object Params {

    val FlowQueryName = "name";
    val FlowQueryStartTime = "start";
    val FlowQueryEndTime = "end";
    val FlowQuerySrcAddress = "src_addr";
    val FlowQueryDstAddress = "dst_addr";
    val FlowQuerySrcPort = "src_port";
    val FlowQueryDstPort = "dst_port";
    val FlowQueryProtocol = "prot";
  }

  val StartedTimeString = "startedTime"
  val EndedTimeString = "endedTime"
  val StatusString = "status"
  val ParamsString = "params"
  val ProgressString = "progress"
  val StatusDetailString = "statusDetail"

  /*def apply(docOption: Option[Document]): Option[Job] = {
    docOption.map { doc =>
      val id: String = doc.getString("_id")
      val `type` = doc.getString("type")
      val name = doc.getString("name")
      val startedTime: Option[String] = Option(doc.getString(StartedTime))
      val endedTime = Option(doc.getString(EndedTime))
      val status = doc.getString(StatusString)
      val params = doc.get(ParamsString).map { paramsValue =>
        val m = paramsValue.asDocument().asScala
        m.map { case (k, v: BsonValue) => (k, v.asString().getValue) }.toMap
      }

      val progress = doc.getInteger(Progress)
      val statusDetail = Option(doc.getString(StatusDetail))
      new Job(id, `type`, name, status, startedTime, endedTime, params, progress, statusDetail)
    }
  }

  implicit class JobConverter(job: Job) {
    def toDoc(): Document = {
      var doc = Document("_id" -> job.id, "type" -> job.`type`, "name" -> job.name, StatusString -> job.status,
        Progress -> job.progress)
      job.startedTime.foreach(st => doc += (StartedTime -> st))
      job.endedTime.foreach(et => doc += (EndedTime -> et))
      job.params.foreach(p => doc += (ParamsString -> Document(p)))
      job.statusDetail.foreach(sd => doc += (StatusDetail -> sd))
      doc
    }
  }*/

}

File: application/src/main/java/com/hxm/earlgrey/control/Job.java
package com.hxm.earlgrey.control;


import java.util.Map;
import org.springframework.data.mongodb.core.mapping.Document;

/**
 * Created by hxm on 17-2-13.
 */
@Document(collection = "Jobs")
public class Job {
    public enum TYPE {Test, FlowQuery}

    public enum STATUS {Created, Started, End, Failed, Error, Canceled}

    public static final String ID_DATE_FORMAT = "yyyyMMddHHmmss";
    public static final String DATE_FORMAT = "yyyy-MM-dd-HH:mm:ss";

    public static String JAVA_COMMAND = "java";

    public static final String PARAM_NAME = "name";
    public static final String PARAM_TYPE = "type";
    public static final String PARAM_RUNNING_TIME = "running_time";
    public static final String PARAM_STARTTIME = "start";
    public static final String PARAM_ENDTIME = "end";
    public static final String PARAM_SRCADDRESS = "src_addr";
    public static final String PARAM_DSTADDRESS = "dst_addr";
    public static final String PARAM_SRCPORT = "src_port";
    public static final String PARAM_DSTPORT = "dst_port";
    public static final String PARAM_PROTOCOL = "prot";
    public static final String PARAM_TOS = "tos";


    private String id;
    private TYPE type;
    private String name;
    private String startedTime;
    private String endedTime;
    private Map<String, String> params;
    private STATUS status = STATUS.Created;
    private int progress;
    private String statusDetail;


    public Job() {
    }

    public Job(String id, TYPE type, String name, Map<String, String> params) {
        this.id = id;
        this.type = type;
        this.name = name;
        this.params = params;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void setType(TYPE type) {
        this.type = type;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Map<String, String> getParams() {
        return params;
    }

    public void setParams(Map<String, String> params) {
        this.params = params;
    }

    public TYPE getType() {
        return type;
    }

    public String getName() {
        return name;
    }

    public String getStartedTime() {
        return startedTime;
    }

    public void setStartTime(String startedTime) {
        this.startedTime = startedTime;
    }

    public String getEndedTime() {
        return endedTime;
    }

    public void setEndedTime(String endedTime) {
        this.endedTime = endedTime;
    }

    public STATUS getStatus() {
        return status;
    }

    public void setStatus(STATUS status) {
        this.status = status;
    }

    public int getProgress() {
        return progress;
    }

    public void setProgress(int progress) {
        this.progress = progress;
    }

    public String getStatusDetail() {
        return statusDetail;
    }

    public void setStatusDetail(String statusDetail) {
        this.statusDetail = statusDetail;
    }
}
